awx/ui_next/src/screens/Host/HostList/HostList.jsx

```diff
@@ -0,0 +1,273 @@
+  ToolbarAddButton,
+  ToolbarDeleteButton,
+} from '@components/PaginatedDataList';
+import { getQSConfig, parseQueryString } from '@util/qs';
+
+import HostListItem from './HostListItem';
+
+const QS_CONFIG = getQSConfig('host', {
+  page: 1,
+  page_size: 20,
+  order_by: 'name',
+});
+
+class HostsList extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      hasContentLoading: true,
+      contentError: null,
+      deletionError: null,
+      hosts: [],
+      selected: [],
+      itemCount: 0,
+      actions: null,
+      toggleError: false,
+      toggleLoading: false,
+    };
+
+    this.handleSelectAll = this.handleSelectAll.bind(this);
+    this.handleSelect = this.handleSelect.bind(this);
+    this.handleHostDelete = this.handleHostDelete.bind(this);
+    this.handleDeleteErrorClose = this.handleDeleteErrorClose.bind(this);
+    this.loadHosts = this.loadHosts.bind(this);
+    this.handleHostToggle = this.handleHostToggle.bind(this);
+    this.handleHostToggleErrorClose = this.handleHostToggleErrorClose.bind(
+      this
+    );
+  }
+
+  componentDidMount() {
+    this.loadHosts();
+  }
+
+  componentDidUpdate(prevProps) {
+    const { location } = this.props;
+    if (location !== prevProps.location) {
+      this.loadHosts();
+    }
+  }
+
+  handleSelectAll(isSelected) {
+    const { hosts } = this.state;
+
+    const selected = isSelected ? [...hosts] : [];
+    this.setState({ selected });
+  }
+
+  handleSelect(row) {
+    const { selected } = this.state;
+
+    if (selected.some(s => s.id === row.id)) {
+      this.setState({ selected: selected.filter(s => s.id !== row.id) });
+    } else {
+      this.setState({ selected: selected.concat(row) });
+    }
+  }
+
+  handleDeleteErrorClose() {
+    this.setState({ deletionError: null });
+  }
+
+  handleHostToggleErrorClose() {
+    this.setState({ toggleError: false });
+  }
+
+  async handleHostDelete() {
+    const { selected } = this.state;
+
+    this.setState({ hasContentLoading: true });
+    try {
+      await Promise.all(selected.map(host => HostsAPI.destroy(host.id)));
+    } catch (err) {
+      this.setState({ deletionError: err });
+    } finally {
+      await this.loadHosts();
+    }
+  }
+
+  async handleHostToggle(hostToToggle) {
+    const { hosts } = this.state;
+    this.setState({ toggleLoading: true });
+    try {
+      const { data: updatedHost } = await HostsAPI.update(hostToToggle.id, {
+        enabled: !hostToToggle.enabled,
+      });
+      this.setState({
+        hosts: hosts.map(host =>
+          host.id === updatedHost.id ? updatedHost : host
+        ),
+      });
+    } catch (err) {
+      this.setState({ toggleError: true });
+    } finally {
+      this.setState({ toggleLoading: false });
+    }
+  }
+
+  async loadHosts() {
+    const { location } = this.props;
+    const { actions: cachedActions } = this.state;
+    const params = parseQueryString(QS_CONFIG, location.search);
+
+    let optionsPromise;
```

Comment:
A little convoluted here avoiding re-fetching actions... Maybe delegate some of this to a `this.loadActions()` that returns `this.state.actions` if set or fetches if not and extracts them out of the `data` key before returning. Then you could do something like:

`const promises = Promise.all([HostsAPI.read(params), this.loadActions()]);`